feat: Add OpenTelemetry support for Subscribe Side

mukund-ananthu opened the pull request:
Subscribe Side OpenTelemetry Code

mukund-ananthu committed:
Add SubscriberOptions and OpenTelemetry Subscriber option

mukund-ananthu committed:
Set / Override OpenTelemetry Subscriber settings based on Python Version

mukund-ananthu committed:
Add OpenTelemetryContextGetter used for trace propagation

mukund-ananthu committed:
Add Subscribe Span and SubscribeOpenTelemetry class

mukund-ananthu committed:
Add receipt modack start and end events to subscribe span

mukund-ananthu committed:
Modify SubscriberMessage to include opentelemetry data and include
SubscribeSpan information in the SubscriberMessage's Open Telemetry data

mukund-ananthu committed:
Add ack, modack, nack start events to subscribe span

mukund-ananthu committed:
Add Open Telemetry data to Ack, Modack, Nack request objects

* This is required for the Open Telemetry information to be percolated
  to the dispatcher which performs acks, nacks and modacks async.
* The Open Telemetry data passed into the dispatcher will be used to add
  events such as ack, modack, nack end to the subscribe span

mukund-ananthu committed:
Add ack result, ack end event to subscribe span and end it

mukund-ananthu committed:
Add tests for ack and ack_with_response methods of Subscriber Message

* Both these methods should add the "ack start" event to the subscribe
  span

mukund-ananthu committed:
Add tests for nack and nack_with_response methods of Subscriber Message

* Both these methods should add "nack start" events to the subscribe
  span

mukund-ananthu committed:
Add tests for modify_ack_deadline and modify_ack_deadline_with_response

* Both these methods should add the "modack start" event to the
  subscribe span

mukund-ananthu committed:
Add tests for set_subscribe_span_result method of Subscribe
OpenTelemetry

mukund-ananthu committed:
Add test for dispatcher ack method

mukund-ananthu committed:
Add test for retry_acks method in dispatcher

mukund-ananthu committed:
Pass the OpenTelemetry data to the dispatcher when message is acked,
nacked or modacked

mukund-ananthu committed:
Record nack and modack end events, set their results and end them in
dispatcher

mukund-ananthu committed:
Record modack, nack ends after successful retries

mukund-ananthu committed:
Fix mypy errors

mukund-ananthu committed:
Add modack events for lease management and receipt modack

mukund-ananthu committed:
Add drop event to subscribe span

mukund-ananthu committed:
Fix mypy errors

mukund-ananthu committed:
Add subscriber concurrency control span

mukund-ananthu committed:
Add subscribe scheduler span

mukund-ananthu committed:
End subscribe scheduler span

mukund-ananthu committed:
Add subscription_id property to SubscribeOpenTelemetry

* This field will then be used for span names of spans like Process Span

mukund-ananthu committed:
Start process span

mukund-ananthu committed:
Add end process span to SubcribeOpenTelemetry

mukund-ananthu committed:
End process span when ack and nack occur

mukund-ananthu committed:
End process span and add event when messages are dropped

hongalex commented on google/cloud/pubsub_v1/subscriber/message.py:
```diff
@@ -252,6 +268,10 @@ def ack(self) -> None:
         """
+        if self.opentelemetry_data:
+            self.opentelemetry_data.add_subscribe_span_event("ack start")
+            self.opentelemetry_data.add_process_span_event("ack called")
```
This span should be added when the user calls msg.Ack(), rather than when the internal library ack is called.

hongalex replied:
Ok reading this, it's the subscribe span event that is actually in the wrong place. That should start with the RPC is invoked, not when the user calls msg.Ack.

mukund-ananthu replied:
Got it. Moved the `add_subscribe_span_event("ack start")` from message.py to dispatcher.py(where the actual RPC call is made). Did this for both ack and ack_with response. Similarly for nack / nack_with_response and modify_ack_deadline and modify_ack_deadline_with_response

hongalex commented on google/cloud/pubsub_v1/subscriber/_protocol/dispatcher.py:
```diff
@@ -303,6 +316,10 @@ def drop(
         """
+        for item in items:
+            if item.opentelemetry_data:
+                item.opentelemetry_data.set_subscribe_span_result("dropped")
```
I think we call this "expired" in the design, but if you prefer dropped to match the client library method, that's fine.

mukund-ananthu replied:
Sure, renamed to "expired"

hongalex commented on google/cloud/pubsub_v1/subscriber/message.py:
```diff
@@ -429,12 +467,17 @@ def nack(self) -> None:
         or another.
         """
+        if self.opentelemetry_data:
+            self.opentelemetry_data.add_subscribe_span_event("nack start")
```
Same comment as ack, event should be added when the RPC is invoked

mukund-ananthu replied:
Done. Response https://github.com/googleapis/python-pubsub/pull/1252#discussion_r1769706317

hongalex commented on google/cloud/pubsub_v1/subscriber/message.py:
```diff
@@ -406,6 +439,8 @@ def modify_ack_deadline_with_response(self, seconds: int) -> "futures.Future":
 
         """
+        if self.opentelemetry_data:
+            self.opentelemetry_data.add_subscribe_span_event("modack start")
```
Same as above: make this event when the RPC is called.

mukund-ananthu replied:
Done. Response: https://github.com/googleapis/python-pubsub/pull/1252#discussion_r1769706317

hongalex commented on google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:
```diff
@@ -1110,6 +1150,7 @@ def _on_response(self, response: gapic_types.StreamingPullResponse) -> None:
                 )
                 return
 
+            i: int = 0
             for received_message in received_messages:
```
Not a python expert, but does using enumerate work better here instead of incrementing the counter manually?

mukund-ananthu replied:
Reason for not trying to enumerate here is that `subscribe_opentelemetry[]` can either be having the same length as the received_messages list when opentelemetry is enabled, in which case an enumeration can be used, but `subscribe_opentelemetry` is an empty list when opentelemetry is disabled. In this case, enumeration would not work, because received_messages and subscribe_opentelemetry would have N and zero length, and something along the lines of:
`for x, y in zip(received_messages, subscribe_opentelemetry)` would cause issues. Hence, I used just a normal counter here.

mukund-ananthu committed:
Fix the trace.Link from process span to publish span

mukund-ananthu committed:
Add null check before checking if Subscriber message has opentelemetry
data

mukund-ananthu committed:
Add tests to verify process span links and subscribe span parent

mukund-ananthu committed:
Move adding dropped event and ending subscriber span from
dispatcher.drop to leaser.maintain_leases

* dispatcher.drop() is also called by ack, _retry_ack, nack and
  maintain_leases(when it wants to drop expired messages.
* We only want to record the dropped event when the message is expired
  and hence dropped by the leaser

mukund-ananthu committed:
Move ack called subscribe span event from message.ack() to
dispatcher.ack() where the ack RPC is made

mukund-ananthu committed:
Move nack start event from message.nack and message.nack_with_response
to dispatcher.modify_ack_deadline(deadline=0) when nack RPC is made

mukund-ananthu committed:
Move modack start event from message.modify_ack_deadline/with response
to dispatcher.modify_ack_deadline()

mukund-ananthu committed:
Rename the "dropped" event with "expired"

mukund-ananthu committed:
Add modack span

mukund-ananthu committed:
Add is_receipt_modack attribute to modack span

mukund-ananthu committed:
Add ack span

mukund-ananthu committed:
Add ack span for dispatcher._retry_ack

mukund-ananthu committed:
Add nack span

mukund-ananthu committed:
Add nack span for dispatcher._retry_modacks

mukund-ananthu committed:
Fix modacks

* Currently, subscribe_span is ended after modack end event is added to
  the subscribe_span. This should not happen, since subscribe_span
  should be ended only after ack,nack or drop, but not modack
* Remove an extraneous modack start that was present in
  streaming_pull_manager.send_lease_modacks(). This is not required,
  since modack start and end events will anyways be added to the
  subscribe span in the dispatcher.modify_ack_deadline() method.
* Replace the NamedTuple._replace() methods with creation of new
  NamedTuple ModackRequest items. This is because NamedTuple._replace()
  returns a new NamedTuple, rather than replacing the existing one. This
  results in the opentelemetry data not being plumbed down into the
  dispatcher.modify_ack_deadline_method.

mukund-ananthu committed:
Fix mypy errors

mukund-ananthu committed:
Add links to subscribe span of the sampled ack_spans in dispatcher.ack

mukund-ananthu committed:
Add links to ack_span in the subscribe_spans in dispatcher.retry_ack()

mukund-ananthu committed:
Add links to nack spans in the subscribe span in
dispatcher.modify_ack_deadlin()

mukund-ananthu committed:
Add links to nack_spans in subscribe_span in dispatcher.retry_nacks()

mukund-ananthu committed:
Add attributes to the links from subscribe_span to ack_span

mukund-ananthu committed:
Add attributes to links in subscribe_spans to nack_spans

mukund-ananthu committed:
Add links to modack spans in subscribe span

mukund-ananthu commented:
@parthea PTAL

mukund-ananthu committed:
Create nack span even if there are no sampled subscribe span

hongalex commented on google/cloud/pubsub_v1/subscriber/_protocol/streaming_pull_manager.py:
```diff
@@ -1007,22 +1020,84 @@ def _get_initial_request(
+        if self._client.open_telemetry_enabled:
+            subscribe_span_links: List[trace.Link] = []
+            subscribe_spans: List[trace.Span] = []
+            assert len(self._subscription.split("/")) == 4
```
nit: split this just once and save the result

mukund-ananthu replied:
Done.

hongalex commented on google/cloud/pubsub_v1/open_telemetry/subscribe_opentelemetry.py:
```diff
@@ -0,0 +1,279 @@
+            getter=OpenTelemetryContextGetter(),
+        )
+        self._publisher_create_span_context = parent_span_context
+        assert len(subscription.split("/")) == 4
```
nit: same as above, save split result

mukund-ananthu replied:
Done

mukund-ananthu committed:
Add a note stating that OpenTelemetry traces are subject to change

mukund-ananthu committed:
Split the subscription name at one place

* Addressing the review comment
